static/js/typing.js

```diff
@@ -0,0 +1,107 @@
+var typing = (function () {
+var exports = {};
+// How long before we assume a client has gone away
+// and expire its typing status
+var TYPING_STARTED_EXPIRY_PERIOD = 15000; // 15s
+// How frequently 'still typing' notifications are sent
+// to extend the expiry
+var TYPING_STARTED_SEND_FREQUENCY = 10000; // 10s
+// How long after someone stops editing in the compose box
+// do we send a 'stopped typing' notification
+var TYPING_STOPPED_WAIT_PERIOD = 5000; // 5s
+
+var users_currently_typing = [];
+
+function send_typing_notification_ajax(recipients, operation) {
+    channel.post({
+        url: '/json/typing',
+        data: {
+            to: recipients,
+            op: operation
+        },
+        success: function () {},
+        error: function (xhr, error_type) {
+            var response = channel.xhr_error_message("Error sending typing notification", xhr);
+        }
+    });
+}
+
+function check_and_send(operation) {
+    if (operation === 'start') {
+        if (compose.recipient() && compose.message_content()) {
+            send_typing_notification_ajax(compose.recipient(), operation);
+        }
+    } else if (operation === 'stop') {
+        if (compose.recipient()) {
+            send_typing_notification_ajax(compose.recipient(), operation);
+        }
+    }
```

Comment:
I don't understand the stop logic here -- why does it matter if the compose box is open?